tensorflow/compiler/xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -53,6 +53,213 @@ static auto& autotune_cache ABSL_GUARDED_BY(autotune_cache_mu) =
 static int64_t cache_hits ABSL_GUARDED_BY(autotune_cache_mu) = 0;
 static int64_t cache_misses ABSL_GUARDED_BY(autotune_cache_mu) = 0;
 
+GpuGemmConfig GetGpuGemmConfig(const HloInstruction* gemm) {
+  GpuGemmConfig config;
+  config.output_shape = gemm->shape();
+  config.lhs_shape = gemm->operand(0)->shape();
+  config.rhs_shape = gemm->operand(1)->shape();
+  config.backend_config =
+      gemm->backend_config<GemmBackendConfig>().ValueOrDie();
+  config.use_cublaslt =
+      gemm->GetModule()->config().debug_options().xla_gpu_enable_cublaslt();
+  return config;
+}
+
+StatusOr<tensorflow::DataType> EncodePrimitiveTypeAsDataType(
+    PrimitiveType type) {
+  switch (type) {
+    case PRED:
+      return tensorflow::DT_BOOL;
+    case BF16:
+      return tensorflow::DT_BFLOAT16;
+    case F16:
+      return tensorflow::DT_HALF;
+    case F32:
+      return tensorflow::DT_FLOAT;
+    case F64:
+      return tensorflow::DT_DOUBLE;
+    case C64:
+      return tensorflow::DT_COMPLEX64;
+    case C128:
+      return tensorflow::DT_COMPLEX128;
+    case S8:
+      return tensorflow::DT_INT8;
+    case S16:
+      return tensorflow::DT_INT16;
+    case S32:
+      return tensorflow::DT_INT32;
+    case S64:
+      return tensorflow::DT_INT64;
+    case U8:
+      return tensorflow::DT_UINT8;
+    case U16:
+      return tensorflow::DT_UINT16;
+    case U32:
+      return tensorflow::DT_UINT32;
+    case U64:
+      return tensorflow::DT_UINT64;
+    default:
+      return InternalError("Unsupported type in EncodePrimitiveAsDataType.");
+  }
+}
+
+Status DoBlasPlansAutotune(se::Stream* stream, const HloInstruction* instr,
+                           se::DeviceMemoryAllocator* allocator,
+                           const GemmBackendConfig& gemm_config) {
+  GpuGemmConfig config = GetGpuGemmConfig(instr);
+  int64_t batch_size = gemm_config.batch_size();
+
+  const HloModuleConfig& hlo_module_config = instr->GetModule()->config();
+  const int32_t cublas_autotune_level =
+      hlo_module_config.debug_options().xla_gpu_autotune_level();
+  const bool init_cublas_data = cublas_autotune_level >= 2;
+  const bool reinit_cublas_data = cublas_autotune_level >= 3;
+  const bool check_cublas = cublas_autotune_level >= 4;
+  const bool crash_on_checking_failure =
+      hlo_module_config.debug_options()
+          .xla_gpu_crash_on_verification_failures();
+
+  se::RedzoneAllocator input_output_allocator(
+      stream, allocator,
+      PtxOptsFromDebugOptions(hlo_module_config.debug_options()),
+      /*memory_limit=*/std::numeric_limits<int64_t>::max());
```

Comment:
In `DoUncachedGemmAutotune`, the `RedzoneAllocator` is constructed with a given `redzone_size`. Does it make sense to not do that here (and use the default 16MiB)? This might be the primary issue with the previous PR.

Reply:
This could be changed. Note that `DoBlasPlansAutotune` is only called when the XLA flag has been set, though.

Reply:
The previous PR did this too. In that case it was not protected by the flag. Was that change intentional?

Reply:
Yes, that change was intentional.

Reply:
You say "this could be changed." You also say the change is intentional. Can you explain please? I dont quite understand.

Reply:
Moving the instantiation of `se::RedzoneAllocator` from `DoGemmAutotune` to `DoBlasPlansAutotune` was intentional. A value for `redzone_size` could be prescribed to avoid using the default.